packages/vite/src/node/server/middlewares/transform.ts

```diff
@@ -122,10 +122,15 @@ export function transformMiddleware(
 
     let url: string
     try {
-      url = decodeURI(removeTimestampQuery(req.url!)).replace(
-        NULL_BYTE_PLACEHOLDER,
-        '\0',
-      )
+      url = removeTimestampQuery(req.url!)
+        .replace(/(%[A-Za-z0-9]{2})+/g, (match) => {
+          try {
+            return decodeURIComponent(match)
+          } catch {
+            return match
+          }
+        })
+        .replace(NULL_BYTE_PLACEHOLDER, '\0')
     } catch (e) {
       return next(e)
     }
```

Comment:
I think plugins won't expect the string to contain invalid utf8 characters. So instead of trying to keep the invalid escape as-is, I think it's better to skip the whole transform middleware with a code like:
```js
} catch (e) {
  if (isURIMalformedError(e) {
    server.config.logger.warn(colors.yellow("warn about it"))
    return next()
  }
  return next(e)
}
```